src/core/masking.ts

```diff
@@ -138,10 +199,176 @@ function maskSelectionSet(
+) {
+  if (Array.isArray(data)) {
+    return data.map((item, index): unknown => {
+      return addFieldAccessorWarnings(
+        memo[index] ?? Object.create(null),
+        item,
+        selectionSetNode,
+        [...path, index],
+        context
+      );
+    });
+  }
+
+  return selectionSetNode.selections.reduce<any>((memo, selection) => {
+    switch (selection.kind) {
+      case Kind.FIELD: {
+        const keyName = resultKeyNameFromField(selection);
+        const childSelectionSet = selection.selectionSet;
+
+        if (keyName in memo) {
+          return memo;
+        }
+
+        let value = data[keyName];
+
+        if (childSelectionSet) {
+          value = addFieldAccessorWarnings(
+            memo[keyName] ?? Object.create(null),
+            data[keyName] as Record<string, unknown>,
+            childSelectionSet,
+            [...path, keyName],
+            context
+          );
+        }
+
+        if (__DEV__) {
+          addAccessorWarning(memo, value, keyName, path, context);
+        }
+
+        if (!__DEV__) {
+          memo[keyName] = data[keyName];
+        }
+
+        return memo;
+      }
+      case Kind.INLINE_FRAGMENT: {
+        return addFieldAccessorWarnings(
+          memo,
+          data,
+          selection.selectionSet,
+          path,
+          context
+        );
+      }
+      case Kind.FRAGMENT_SPREAD: {
+        const fragment = context.fragmentMap[selection.name.value];
+        const mode = getFragmentMaskMode(selection);
+
+        if (mode === "mask") {
+          return memo;
+        }
+
+        if (mode === "unmask") {
+          const [fragmentData] = maskSelectionSet(
+            data,
+            fragment.selectionSet,
+            context,
+            path
+          );
+
+          return Object.assign(memo, fragmentData);
+        }
+
+        return addFieldAccessorWarnings(
+          memo,
+          data,
+          fragment.selectionSet,
+          path,
+          context
+        );
+      }
+    }
+  }, memo);
+}
+
+function addAccessorWarning(
+  data: Record<string, any>,
+  value: any,
+  fieldName: string,
+  path: PathSelection,
+  context: MaskingContext
+) {
+  let currentValue = value;
+  let warned = false;
+
+  Object.defineProperty(data, fieldName, {
+    get() {
+      if (context.disableWarnings) {
+        return currentValue;
+      }
+
+      if (!warned) {
+        invariant.warn(
+          "Accessing unmasked field on %s at path '%s'. This field will not be available when masking is enabled. Please read the field from the fragment instead.",
+          context.operationName ?
+            `${context.operationType} '${context.operationName}'`
+          : `anonymous ${context.operationType}`,
+          getPathString([...path, fieldName])
+        );
+        warned = true;
+      }
+
+      return currentValue;
+    },
+    set(value) {
+      currentValue = value;
+    },
+    enumerable: true,
+    configurable: true,
+  });
+}
```

Comment:
```suggestion
) {
  Object.defineProperty(data, fieldName, {
    get() {
      if (context.disableWarnings) {
        return value;
      }

      invariant.warn(
        "Accessing unmasked field on %s at path '%s'. This field will not be available when masking is enabled. Please read the field from the fragment instead.",
        context.operationName ?
          `${context.operationType} '${context.operationName}'`
        : `anonymous ${context.operationType}`,
        getPathString([...path, fieldName])
      );

      // this will call the setter and replace this with a normal writable property
      return (data.value = value);
    },
    set(value) {
      Object.defineProperty(data, fieldName, { writable: true, value });
    },
    enumerable: true,
    configurable: true,
  });
}
```

This way we wouldn't need to track `currentValue` and `warned` - if my assumption that any `set` call should disable warnings is correct. (This should also save us that check for the presence of a descriptor and `delete`.)